README.md
<img src="img/vanilla_ISE_logo.png">

##### Your endpoint specialists will thank you later...
[![published](https://static.production.devnetcloud.com/codeexchange/assets/images/devnet-published.svg)](https://developer.cisco.com/codeexchange/github/repo/obrigg/Vanilla-ISE)
## The Challenge

Cisco ISE (Identity Services Engine) is **THE** policy engine for your network. It provides many (many) different services that are all required to meet today's user expectations while protecting the organization from threats:
* TACACS+/RADIUS for central management of networking equipment.
* Network Admission Control (NAC) identifying endpoints as they connect to the network (both wired and wireless).
* Group-based Policy using Cisco TrustSec.
* VPN policy with Cisco ASA/FTD.
* Guest lifecycle management.
* Device profiling.
* And more...

While Cisco ISE's richness of features and capabilities is highly appreciated by networking and security experts, endpoint technicians and helpdesk representatives prefer a simpler, limited GUI for their day-to-day tasks.

A specific capability asked by endpoint teams is the concept of a "voucher": a time-limited access token, given to a specific endpoint in order to grant immediate network access (while bypassing security checks) for a distressed user, allowing the technicians to remediate gaps in the endpoint's posture later.

## The Solution

Using Cisco ISE's open APIs, primarily ERS (External RESTful Services), we have created "Vanilla ISE": a simple UI for endpoint technicians and helpdesk representatives.

### Vanilla ISE's Capabilities:
* List the network access devices configured on ISE.
<img src="img/device list.png">
* Switch view provides a graphical representation of the switch. Which ports are up/down, authenticated or unauthenticated. Clicking a port will provide more details about the endpoint behind the port (similar to device query - but shorter response time).
  * <span style="color:green">Green</span>: Authenticated endpoint(s).
  * <span style="color:red">Red</span>: Unauthenticated endpoint(s).
  * <span style="color:grey">Grey</span>: The interface is not connected.
  
  <img src="img/switch_view.png">
  <img src="img/switch_view_port.png">

* The switch view enables new capabilities:
  * Clear port sessions: issues the `clear authentication session interface <interface>` command on the interface, clearing existing authentication sessions (quicker and less disruptive than shut/no shut)
  * Bypass port: (AKA port voucher) temporarily removes the dot1x configuration from the interface for 24 hours. The bypassed interface will appear in the vouchers page and can be revoked from that page. 
  
  <img src="img/switch_view_port_menu.png">

* Query a network device for current authentication sessions.
<img src="img/device query.png">

* Query a specific endpoint for details (status, NAD's IP and interface, authentication mechanism, username, and failure reason - if failed)
<img src="img/endpoint query.png">

* Grant and revoke network access "vouchers".
<img src="img/voucher list.png">
<img src="img/add voucher.png">

* **Vanilla ISE now requires authentication**. Use the your ISE credentials to authenticate to Vanilla ISE.

* Auditing: Vanilla ISE will keep track of the users creating and revoking vouchers, as well as send audit messages to a syslog server configured in the environment variables.

##### Communication with the devices, and data parsing is powered by <img src="/img/pyats.png">
More information about pyATS is available at: https://developer.cisco.com/pyats/
## Running vanilla ISE:

There are several options for running vanilla ISE:
1. Running the code on a computer/server with Python.
2. Running the code on a Docker container. Requires to <a href="https://docs.docker.com/get-docker/"> install Docker</a>.

### Enable ISE ERS API

The ISE REST APIs (AKA External RESTful Services or ERS) are disabled by default for security. You must enable it:
1. Login to your ISE PAN using the admin or other SuperAdmin user.
2. Navigate to **Administration** > **System** > **Settings** and select **ERS Settings** from the left panel.
4. Enable the ERS APIs by selecting **Enable ERS for Read/Write**
5. Do not enable CSRF unless you know how to use the tokens.
6. Select **Save** to save your changes.

Note: its good practice to disable CSRF to make sure you are able to authenticate successfully.

<a href="https://community.cisco.com/t5/security-documents/ise-ers-api-examples/ta-p/3622623#toc-hId--623796905"> Reference to official documentation </a>

### Option 1: Running Vanilla ISE on a server/workstation
#### Virtual Environment

I recommend running Vanilla ISE in a Python virtual environment. This will help keep your host system clean and allow you to have multiple environments to try new things. If you are not using a virtual environment, start at the download/clone step below.

You will also need Python 3 and venv installed on your host system.

In your project directory, create your virtual environment
``` console
python3 -m venv env
```
Activate the new virtual environment:
``` console
source env/bin/activate
```
Download or clone the Vanilla ISE repository:

``` console
git clone https://github.com/obrigg/Vanilla-ISE.git
```
Install the required packages
```
cd Vanilla-ISE
pip install -r requirements.txt
```
#### Set environment variables (one-time, non-persistent)
```
export ISE_IP= <ISE hostname/IP>
export ISE_USER= <ISE username>
export ISE_PASSWORD= <ISE password>
export SYSLOG_SERVER= <Syslog server IP, for auditing>
export SWITCH_USER= <username for network devices>
export SWITCH_PASS= <password for network devices>
export SWITCH_ENABLE= <enable password for network devices>
```
#### Set environment variables (persistent, for multiple activations)
Edit the env/bin/activate file with nano/vi/other editor
```
nano env/bin/activate
```
Add the following lines to the file, to make the environment variables persist for multiple activations of the environment.
```
export ISE_IP= <ISE hostname/IP>
export ISE_USER= <ISE username>
export ISE_PASSWORD= <ISE password>
export SYSLOG_SERVER= <Syslog server IP, for auditing>
export SWITCH_USER= <username for network devices>
export SWITCH_PASS= <password for network devices>
export SWITCH_ENABLE= <enable password for network devices>
```
#### Run the code
```
python app.py
```

### Option 2: Running Vanilla ISE as a Docker container (CLI)
#### Create an environment variable file
```
ISE_IP= <ISE hostname/IP>
ISE_USER= <ISE username>
ISE_PASSWORD= <ISE password>
SYSLOG_SERVER= <Syslog server IP, for auditing>
SWITCH_USER= <username for network devices>
SWITCH_PASS= <password for network devices>
SWITCH_ENABLE= <enable password for network devices>
```

##### Run the Docker container as a daemon:
`docker run -d --env-file <path to env file> -p 5000:5000 -v <path to data dir>:/Vanilla-ISE/data obrigg/vanilla-ise`

##### Run the Docker container in interactive mode:
`docker run -ti --env-file <path to env file> -p 5000:5000 -v <path to data dir>:/Vanilla-ISE/data obrigg/vanilla-ise`

### Option 3: Running Vanilla ISE as a Docker container (Docker Compose)
#### Create a `docker-compose.yaml` file (sample in repo)

```
version: '3'
services:
  vanilla-ise:
    image: obrigg/vanilla-ise
    container_name: vanilla-ise
    environment:
      - ISE_IP= <ISE hostname/IP>
      - ISE_USER= <ISE username>
      - ISE_PASSWORD= <ISE password>
      - SYSLOG_SERVER= <Syslog server IP, for auditing>
      - SWITCH_USER= <username for network devices>
      - SWITCH_PASS= <password for network devices>
      - SWITCH_ENABLE= <enable password for network devices>
    volumes:
      - <path to data dir>:/Vanilla-ISE/data
    ports:
      - 5000:5000
```

#### Run the Docker container:
`docker-compose up -d`

----

You can access the Vanilla ISE GUI at `http://<host ip>:5000` using your ISE username/password to login.

----
### Licensing info
Copyright (c) 2021 Cisco and/or its affiliates.

This software is licensed to you under the terms of the Cisco Sample
Code License, Version 1.1 (the "License"). You may obtain a copy of the
License at

               https://developer.cisco.com/docs/licenses

All use of the material herein must be in accordance with the terms of
the License. All rights not expressly granted by the License are
reserved. Unless required by applicable law or agreed to separately in
writing, software distributed under the License is distributed on an "AS
IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied.


changelog.md
# Change log

##### June-2022
* Fetching NAD IDs instead of IPs for faster retrieval.
* Added a graceful disconnect from the switches, in order to avoid utilizing all VTY sessions.
* Black'd, for PEP8.
* Gracefully handling unknown OUI.
* Added a "logout" button.
* Device list uses device ID instead of IP addresses (for improved performance)

##### 05-May-2022
* Added a page with a graphical representation of a switch and its port statuses.
* For each port, a detail page with the session information for the specific port was added.
* Ability to clear the authentication sessions for specific interfaces - by right clicking a port in the switch view, and choosing "Clear port".
* Ability to create a "port voucher", temporarily disabling dot1x on that interface for 24 hours.

##### 23-February-2022
* Documentation updates/clarifications, thanks to Roddie.
* Sample Docker-Compose file, thanks to Roddie.

##### 5-July-2021
* NAD List fetch converted to async for faster response.

##### 12-May-2021
* Added a welcome page, to avoid loading delay for networks with many NADs.
* Added jquery as a local file.
* Changed `show authentication session` to `show access-session` command to support older switches.

##### 6-May-2021
* Added search bar for the device list and voucher list.

##### 5-May-2021
* Updated NAD list with pagination, for more than 20 NADs.

##### 1-May-2021
* Added static vlans to device query.

##### 17-April-2021
* Added records of the user creating each voucher.
* Added auditing via syslog.

##### 10-April-2021
* Added colors to the backend.

##### 9-April-2021
* Added user authentication to Vanilla ISE.

##### 4-April-2021
* Added support for multiple voucher groups.



mab_cleanup.py
import os
import requests
import json
import xmltodict
from time import time
from requests.auth import HTTPBasicAuth
from requests.packages.urllib3.exceptions import InsecureRequestWarning
requests.packages.urllib3.disable_warnings(InsecureRequestWarning)


######   Setting up the environment ######
ise_user = os.environ.get('ISE_USER', "admin")
ise_password = os.environ.get('ISE_PASSWORD', "")
base_url = "https://" + os.environ.get('ISE_IP', "") + ":9060/ers/config/"

auth = HTTPBasicAuth(ise_user, ise_password)
headers = {"Content-Type": "application/json",
           "Accept": "application/json"}


def get_ise_cleanup_groups():
    '''
    This function will search ISE for endpoint groups that include the
    tag #cleanup in their description for the cleanup process.
    It will return a list of group IDs.
    '''
    print("\033[0;34mFetching ISE's endpoint groups...\033[0m")
    url = base_url + "endpointgroup?size=100"
    response = requests.get(url=url, auth=auth, headers=headers, verify=False)
    if response.status_code == 200:
        cleanup_groups = []
        groups = response.json()['SearchResult']['resources']
        for group in groups:
            if "#cleanup" in group['description'].lower():
                print(f"\033[0;34mEndpoint group {group['name']} will be cleaned up\033[0m")
                cleanup_groups.append(group['id'])
        return(cleanup_groups)
    else:
        print(f"\033[0;34mERROR: {response.text}\033[0m")
        return("ERROR")


def get_endpoints_by_group_id(groupId: str):
    if groupId == "":
        url = base_url + "endpoint?size=10"
    else:
        url = base_url + f"endpoint?size=100&filter=groupId.EQ.{groupId}"
    isFinished = False
    list_of_endpoints = []
    while not isFinished:
        response = requests.get(url=url, headers=headers, auth=auth, verify=False)
        if response.status_code != 200:
            print(f"\033[0;34mAn error has occurred: {response.json()}\033[0m")
        else:
            endpoints = response.json()['SearchResult']['resources']
            if len(endpoints) > 0:
                list_of_endpoints += endpoints
            if 'nextPage' not in response.json()['SearchResult'].keys():
                isFinished = True
            else:
                url = response.json()['SearchResult']['nextPage']['href']
    if len(list_of_endpoints) > 0:
        return(list_of_endpoints)


def check_ise_auth_status(mac_address: str):
    url = "https://" + os.environ.get('ISE_IP', "") + \
        "/admin/API/mnt/Session/MACAddress/" + mac_address
    response = requests.get(url=url, auth=auth, verify=False)
    if response.status_code != 200:
        print(f"\033[0;34mAn error has occurred: {response.text}\033[0m")
        return("ERROR")
    else:
        status = xmltodict.parse(response.text)
        return(status['sessionParameters']['acct_status_type'])
    

def delete_endpoint(endpoint_id: str):
    url = base_url + "endpoint/" + endpoint_id
    response = requests.delete(url=url, auth=auth, headers=headers, verify=False)
    if response.status_code != 204:
        print(f"\033[0;34mError deleting endpoint {endpoint_id}: {response.text}\033[0m")


def main():
    # Gather relevant cleanup groups
    cleanup_groups = get_ise_cleanup_groups()
    # Gather all relevant endpoints
    list_of_endpoints = []
    for groupId in cleanup_groups:
        endpoints = get_endpoints_by_group_id(groupId)
        if endpoints != None:
            list_of_endpoints += endpoints
    print(f"\033[0;34mReceived a total of {len(list_of_endpoints)} endpoints.\033[0m")

    for endpoint in list_of_endpoints:
        #print(endpoint)
        status = check_ise_auth_status(endpoint['name'])
        if status == "Stop":
            print(f"\033[0;34mDeleting endpoint {endpoint['name']}\033[0m")
            delete_endpoint(endpoint['id'])

if __name__ == "__main__":
    main()

app.py
__version__ = "22.06.26.01"
__author__ = "Oren Brigg & Ramona Renner"
__author_email__ = "obrigg@cisco.com / ramrenne@cisco.com"
__license__ = "Cisco Sample Code License, Version 1.1 - https://developer.cisco.com/site/license/cisco-sample-code-license/"


""" Copyright (c) 2021 Cisco and/or its affiliates.
This software is licensed to you under the terms of the Cisco Sample
Code License, Version 1.1 (the "License"). You may obtain a copy of the
License at
           https://developer.cisco.com/docs/licenses
All use of the material herein must be in accordance with the terms of
the License. All rights not expressly granted by the License are
reserved. Unless required by applicable law or agreed to separately in
writing, software distributed under the License is distributed on an "AS
IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied.
"""

# Import Section
from flask import Flask, render_template, request, url_for, redirect, session
from requests.auth import HTTPBasicAuth
import backend
import requests
from time import ctime, sleep, time
from threading import Thread

# Global Variables
app = Flask(__name__)
app.secret_key = "CiscoISEShared"

# Methods
def convert_voucher_list(voucher_list):
    """
    This function will receive a voucher list with MAC addresses (in "the cisco way" format)
    and expire timestamps (in seconds since the epoch, in UTC) and convert it
    to MAC addresses (in the format xx:xx:xx:xx:xx:xx) and expire dates (as strings in local time).
    """
    for voucher in voucher_list:

        if voucher["type"] == "host":
            voucher["mac"] = normalize_mac_format(voucher["mac"])
            voucher["ExpDate"] = ctime(voucher["duration"])

        elif voucher["type"] == "port":
            voucher["ExpDate"] = ctime(voucher["duration"])

    return voucher_list


def propagate_backend_exception(backend_response):
    """
    This function checks the backend response for ERROR indictors to raise an frontend
    exception and thereby display a customized error message in case of an backend error.
    """
    if "ERROR" in str(backend_response):
        raise Exception(str(backend_response))


def voucher_cleanup_loop():
    while True:
        backend.voucher_cleanup()
        sleep(10 * 60)


def normalize_mac_format(mac_address):
    """
    This function normalizes mac addresses of different formats
    e.g. xxxxxxxxxxxx, xx-xx-xx-xx-xx-xx, xx.xx.xx.xx.xx.xx or xxxx.xxxx.xxxx
    to xx:xx:xx:xx:xx:xx
    """
    # Remove all :, - and .
    mac_address = mac_address.replace(".", "").replace(":", "").replace("-", "")
    # Add : after every second char
    normalized_mac_address = ":".join(
        a + b for a, b in zip(mac_address[::2], mac_address[1::2])
    )

    return normalized_mac_address


# Routes
@app.route("/", methods=["GET"])
def index():
    """
    This function shows an empty welcome page (GET)
    """
    if type(session.get("logged_in")) == int and session.get("logged_in") > int(time()):
        session["logged_in"] = int(time()) + backend.timeout
        print(
            f"Login for {session['username']} extended until: {ctime(session['logged_in'])}"
        )
        try:
            if request.method == "GET":
                return render_template("welcome.html")

        except Exception as e:
            print(e)
            return render_template("endpointQuery.html", error=True, errorcode=e)
    else:
        print("First you need to login")
        return redirect(url_for("login"))


@app.route("/deviceList", methods=["GET", "POST"])
def deviceList():
    """
    This function will show all NAD devices in a table (GET request) and
    redirect to the specific query page of a device (POST request). The redirection
    can be manually triggered by a "Query Device" link in each table row and thereby
    for each NAD device.
    """
    if type(session.get("logged_in")) == int and session.get("logged_in") > int(time()):
        session["logged_in"] = int(time()) + backend.timeout
        print(
            f"Login for {session['username']} extended until: {ctime(session['logged_in'])}"
        )
        try:
            if request.method == "GET":

                device_list = backend.get_all_NADs()
                propagate_backend_exception(device_list)

                return render_template("deviceList.html", device_list=device_list)

        except Exception as e:
            print(e)
            return render_template(
                "deviceList.html", error=True, errorcode=e, reloadlink="/"
            )
    else:
        print("First you need to login")
        return redirect(url_for("login"))


app.route("/deviceQuery?ip_address=<ip_address>")


@app.route("/deviceQuery", methods=["GET", "POST"])
def deviceQuery():
    """
    This function shows an empty page with IP query field and button (GET)
    or a list of queried devices with associated session information (POST).
    """
    if type(session.get("logged_in")) == int and session.get("logged_in") > int(time()):
        session["logged_in"] = int(time()) + backend.timeout
        print(
            f"Login for {session['username']} extended until: {ctime(session['logged_in'])}"
        )
        try:
            if request.method == "GET":
                ip_address = request.args.get("ip_address")

            elif request.method == "POST":
                ip_address = request.form.get("ip_address")

            if ip_address == None:
                ip_address = ""
                relevant_sessions = {}
            else:
                relevant_sessions = backend.get_device_auth_sessions(ip_address)
                propagate_backend_exception(relevant_sessions)

            return render_template(
                "deviceQuery.html",
                ip_address=ip_address,
                relevant_sessions=relevant_sessions,
            )

        except Exception as e:
            print(e)
            return render_template("deviceQuery.html", error=True, errorcode=e)
    else:
        print("First you need to login")
        return redirect(url_for("login"))


app.route("/switchView?ip_address=<ip_address>&success=<success>")
app.route("/switchView?nad_id=<nad_id>&success=<success>")


@app.route("/switchView", methods=["GET", "POST"])
def switchView():
    """
    This function shows an graphical representation of a switch and its port status.
    """
    if type(session.get("logged_in")) == int and session.get("logged_in") > int(time()):
        session["logged_in"] = int(time()) + backend.timeout
        print(
            f"Login for {session['username']} extended until: {ctime(session['logged_in'])}"
        )
        try:
            if request.method == "GET":

                nad_id = request.args.get("nad_id")
                if nad_id == None:
                    ip_address = request.args.get("ip_address")
                else:
                    ip_address = backend.get_nad_ip(nad_id)
                
                print(f"IP: {ip_address}")

            elif request.method == "POST":

                ip_address = request.form.get("ip_address")

            success = request.args.get("success")

            if ip_address == None:
                ip_address = ""
                detailed_switch_status = {}
            else:
                detailed_switch_status = backend.get_device_ports(ip_address)
                propagate_backend_exception(detailed_switch_status)

            return render_template(
                "switchView.html",
                ip_address=ip_address,
                success=success,
                detailed_switch_status=detailed_switch_status,
            )

        except Exception as e:
            print(e)
            return render_template("switchView.html", error=True, errorcode=e)
    else:
        print("First you need to login")
        return redirect(url_for("login"))


app.route("/portAction?ip_address=<ip_address>&interface=<interface>&action=<action>")


@app.route("/portAction", methods=["GET", "POST"])
def portAction():
    """
    This function handles the port actions: clear port sessions and bypass.
    """
    if type(session.get("logged_in")) == int and session.get("logged_in") > int(time()):
        session["logged_in"] = int(time()) + backend.timeout
        print(
            f"Login for {session['username']} extended until: {ctime(session['logged_in'])}"
        )
        try:
            ip_address = request.args.get("ip_address")
            interface = request.args.get("interface")
            action = request.args.get("action")

            if action == "clear":
                response = backend.clear_port_auth_sessions(ip_address, interface)
                propagate_backend_exception(response)
            elif action == "bypass":
                response = backend.add_port_voucher(
                    ip_address, interface, 24, session["username"]
                )
                propagate_backend_exception(response)

            return redirect("/switchView?ip_address=" + ip_address + "&success=True")

        except Exception as e:
            print(e)
            return render_template("switchView.html", error=True, errorcode=e)
    else:
        print("First you need to login")
        return redirect(url_for("login"))


app.route("/switchViewDetail?ip_address=<ip_address>&interface=<interface>")


@app.route("/switchViewDetail", methods=["GET", "POST"])
def switchViewDetail():
    """
    This function shows the associated session information for a specific interface and device.
    """
    if type(session.get("logged_in")) == int and session.get("logged_in") > int(time()):
        session["logged_in"] = int(time()) + backend.timeout
        print(
            f"Login for {session['username']} extended until: {ctime(session['logged_in'])}"
        )

        ip_address = ""
        interface = ""

        try:
            if request.method == "GET":

                ip_address = request.args.get("ip_address")
                interface = request.args.get("interface")

                relevant_sessions = backend.get_port_auth_sessions(
                    ip_address, interface
                )
                propagate_backend_exception(relevant_sessions)

                return render_template(
                    "switchViewDetail.html",
                    ip_address=ip_address,
                    interface=interface,
                    relevant_sessions=relevant_sessions,
                )

        except Exception as e:
            print(e)
            if "has no authentication sessions" in repr(e):
                return render_template(
                    "switchViewDetail.html",
                    error=False,
                    ip_address=ip_address,
                    interface=interface,
                    relevant_sessions={},
                )
            elif "No access sessions on" in repr(e):
                return render_template(
                    "switchViewDetail.html",
                    error=False,
                    no_access_session=True,
                    ip_address=ip_address,
                    interface=interface,
                    relevant_sessions={},
                )

            return render_template("switchViewDetail.html", error=True, errorcode=e)

    else:
        print("First you need to login")
        return redirect(url_for("login"))


@app.route("/voucher", methods=["GET", "POST"])
def voucher():
    """
    This function shows a list of vouchers (GET), allows to add new vouchers
    or revokes existing once (POST). For better user experience the voucher data
    is converted beforehand.
    """
    if type(session.get("logged_in")) == int and session.get("logged_in") > int(time()):
        session["logged_in"] = int(time()) + backend.timeout
        print(
            f"Login for {session['username']} extended until: {ctime(session['logged_in'])}"
        )
        try:
            if request.method == "GET":

                voucher_list = convert_voucher_list(backend.read_voucher_list())
                propagate_backend_exception(voucher_list)

                return render_template(
                    "voucher.html", voucher_list=voucher_list, new_voucher=False
                )

            elif request.method == "POST":
                new_voucher = False
                deleted_voucher = False
                submit_type = request.form.get(
                    "voucher_sumbit"
                )  # Possible values: add, revoke_host, revoke_port

                # Add
                if submit_type == "add":
                    form_mac_address = normalize_mac_format(
                        request.form.get("mac_address_field")
                    )
                    voucher_duration = request.form.get("voucher_duration")
                    voucher_group = request.form.get("voucher_group")

                    response = backend.add_host_voucher(
                        form_mac_address,
                        int(voucher_duration),
                        voucher_group,
                        session["username"],
                    )
                    new_voucher = True

                # Revoke
                elif submit_type == "revoke_host":
                    host_mac = request.form.get("host_mac")
                    response = backend.revoke_host_voucher(
                        host_mac, session["username"]
                    )
                    deleted_voucher = True

                elif submit_type == "revoke_port":
                    port_switch_ip = request.form.get("port_switch_ip")
                    port_interface = request.form.get("port_interface")
                    port_command = request.form.get("port_command")
                    response = backend.revoke_port_voucher(
                        port_switch_ip,
                        port_interface,
                        port_command,
                        session["username"],
                    )
                    deleted_voucher = True

                propagate_backend_exception(response)
                voucher_list = convert_voucher_list(backend.read_voucher_list())
                propagate_backend_exception(voucher_list)

                return render_template(
                    "voucher.html",
                    voucher_list=voucher_list,
                    new_voucher=new_voucher,
                    deleted_voucher=deleted_voucher,
                )

        except Exception as e:
            print(e)
            return render_template("voucher.html", error=True, errorcode=e)
    else:
        print("First you need to login")
        return redirect(url_for("login"))


@app.route("/endpointQuery?mac_address=<mac_address>")
@app.route("/endpointQuery", methods=["GET", "POST"])
def endpointQuery():
    """
    This function shows an empty page with MAC address query field and button (GET)
    or a list of queried endpoint devices with associated auth status (POST).
    """
    if type(session.get("logged_in")) == int and session.get("logged_in") > int(time()):
        session["logged_in"] = int(time()) + backend.timeout
        print(
            f"Login for {session['username']} extended until: {ctime(session['logged_in'])}"
        )
        mac_address = ""
        endpoint_list = ""
        try:
            if request.method == "GET":

                mac_address = request.args.get("mac_address")

            elif request.method == "POST":

                mac_address = request.form.get("mac_address")

            if mac_address == None:
                mac_address = ""
                endpoint_list = {}
            else:
                mac_address = normalize_mac_format(mac_address)
                endpoint_list = backend.check_ise_auth_status(mac_address)
                propagate_backend_exception(endpoint_list)

            return render_template(
                "endpointQuery.html",
                endpoint_list=endpoint_list,
                mac_address=mac_address,
            )

        except Exception as e:
            print(e)
            if "No authentication events for MAC Address" in repr(
                e
            ) and "during the last 24.0 hours." in repr(e):
                return render_template(
                    "endpointQuery.html",
                    error=False,
                    no_events=True,
                    endpoint_list={},
                    mac_address=mac_address,
                )

            return render_template("endpointQuery.html", error=True, errorcode=e)
    else:
        print("First you need to login")
        return redirect(url_for("login"))


@app.route("/login", methods=["GET", "POST"])
def login():
    # Reset an existing sessio
    session["logged_in"] = 0
    """Login Form"""
    if request.method == "GET":
        return render_template("login.html")
    else:
        name = request.form["username"]
        passw = request.form["password"]
        x = backend.initialize_ise(name, passw)
        if x == "Done":
            print("login view - Successful Login")
            session["logged_in"] = int(time()) + backend.timeout
            session["username"] = name
            print(f"Logged in until: {ctime(session['logged_in'])}")
            return redirect(url_for("index"))
        else:
            print("Failed login")
            return render_template("login.html")


# Main Function
if __name__ == "__main__":
    t1 = Thread(target=voucher_cleanup_loop)
    t1.start()
    sleep(1)
    app.run(host="0.0.0.0", debug=False, threaded=True)
